materialize-postgres/driver.go

```diff
@@ -356,90 +348,88 @@ func (t *transactor) addBinding(ctx context.Context, target sql.Table) error {
 	}
 
 	t.bindings = append(t.bindings, b)
+
+	// Create a binding-scoped temporary table for staged keys to load.
+	if _, err := t.load.conn.Exec(ctx, b.createLoadTableSQL); err != nil {
+		return fmt.Errorf("Exec(%s): %w", b.createLoadTableSQL, err)
+	}
+
 	return nil
 }
 
 func (d *transactor) Load(it *pm.LoadIterator, loaded func(int, json.RawMessage) error) error {
 	var ctx = it.Context()
 
-	// We send SQL commands in batches to amortize network round-trips.
-	// All operations of Load run as a single transaction.
-	var batch = newPGBatch()
-	batch.queue(nil, "begin;")
-
-	for _, b := range d.bindings {
-		batch.queue(nil, b.createLoadTableSQL)
-		batch.queue(nil, b.prepLoadInsertSQL)
+	// Use a read-only "load" transaction, which will automatically
+	// truncate the temporary key staging tables on commit.
+	var txn, err = d.load.conn.BeginTx(ctx, pgx.TxOptions{})
+	if err != nil {
+		return fmt.Errorf("DB.BeginTx: %w", err)
 	}
+	defer txn.Rollback(ctx)
 
+	var batch pgx.Batch
 	for it.Next() {
 		var b = d.bindings[it.Binding]
 
 		if converted, err := b.target.ConvertKey(it.Key); err != nil {
 			return fmt.Errorf("converting Load key: %w", err)
 		} else {
-			batch.queueParams(nil, b.execLoadInsertSQL, converted...)
-		}
-
-		if len(batch.buf) > 1024*1024 {
-			if err := batch.roundTrip(ctx, d.load.conn.PgConn()); err != nil {
-				return err
-			}
```

Comment:
Is it okay that we now will create batches larger than 1Mb?

Reply:
That is a good point, I had for some reason assumed that under the hood pgx managed those between the server to keep from building up a huge buffer, but it definitely doesn't. So I'll need to revise this to keep memory usage from blowing up during large txns.